lib/babe/babe.go

```diff
@@ -336,77 +301,73 @@ func (b *Service) initiate() {
 		return
 	}
 
-	err := b.invokeBlockAuthoring()
+	// we should consider better error handling for this - we should
+	// retry to run the engine at some point (maybe the next epoch) if
+	// there's an error.
+	if err := b.runEngine(); err != nil {
+		logger.Criticalf("failed to run block production engine: %s", err)
+	}
+}
+
+func (b *Service) initiateAndGetEpochHandler(ctx context.Context, epoch uint64) (*epochHandler, error) {
+	epochData, err := b.initiateEpoch(epoch)
+	if err != nil {
+		logger.Errorf("failed to initiate epoch %d: %s", epoch, err)
+		return nil, err
+	}
+
+	logger.Debugf("initiated epoch with threshold %s, randomness 0x%x and authorities %v",
+		epochData.threshold, epochData.randomness[:], epochData.authorities)
+
+	epochStartSlot, err := b.epochState.GetStartSlotForEpoch(epoch)
 	if err != nil {
-		logger.Criticalf("block authoring error: %s", err)
+		logger.Errorf("failed to get start slot for current epoch %d: %s", epoch, err)
+		return nil, err
 	}
+
+	return newEpochHandler(ctx,
+		epoch,
+		epochStartSlot,
+		epochData,
+		b.constants,
+		b.handleSlot,
+		b.keypair,
+	)
 }
 
-func (b *Service) invokeBlockAuthoring() error {
+func (b *Service) runEngine() error {
 	epoch, err := b.epochState.GetCurrentEpoch()
 	if err != nil {
 		logger.Errorf("failed to get current epoch: %s", err)
 		return err
 	}
 
 	for {
-		err := b.initiateEpoch(epoch)
-		if err != nil {
-			logger.Errorf("failed to initiate epoch %d: %s", epoch, err)
-			return err
-		}
-
-		logger.Debugf("initiated epoch with threshold %s, randomness 0x%x and authorities %v",
-			b.epochData.threshold, b.epochData.randomness[:], b.epochData.authorities)
-
-		epochStartSlot, err := b.waitForEpochStart(epoch)
-		if err != nil {
-			logger.Errorf("failed to wait for epoch %d start: %s", epoch, err)
-			return err
-		}
-
-		// calculate current slot
-		startSlot := getCurrentSlot(b.slotDuration)
-		intoEpoch := startSlot - epochStartSlot
-
-		// if the calculated amount of slots "into the epoch" is greater than the epoch length,
-		// we've been offline for more than an epoch, and need to sync. pause BABE for now, syncer will
-		// resume it when ready
-		if b.epochLength <= intoEpoch && !b.dev {
-			logger.Debugf(
-				"pausing BABE, need to sync since we have %d slots (start slot %d) into the epoch starting at %d",
-				intoEpoch, startSlot, epochStartSlot)
-			return b.Pause()
-		}
-
-		if b.dev {
-			intoEpoch = intoEpoch % b.epochLength
-		}
-
-		logger.Infof("current epoch %d has %d slots", epoch, intoEpoch)
-
-		// get start slot for current epoch
-		nextEpochStart, err := b.epochState.GetStartSlotForEpoch(epoch + 1)
-		if err != nil {
-			logger.Errorf("failed to get start slot for next epoch %d: %s", epoch+1, err)
-			return err
-		}
+		err = func() error {
+			ctx, cancel := context.WithCancel(b.ctx)
```

Comment:
why do we have these nested contexts? Makes it pretty difficult to read.

Reply:
Not sure what is the problem here.

To me it makes sense to have a nested context to manage a new epoch handler. This way we can manage it independently. if `epochhandler.run` times out in one loop run, we can still move on to other one.

Other option is to use the same context `b.ctx`, but I think that approach has problems. Using `b.ctx.cancel` could impact other things.
( cancel should only be called by function that creates the context. You can pass it around, but it is not recommended since it is likely to create problems.)

Let me know if I might have misunderstood what you have written.

Reply:
yeah, I agree with Kishan, the point is to have a context per epoch that is canceled when the epoch finished

Reply:
I'm of the proponent where we should not be using `context.WithCancel()` within package boundaries.  I know we already do this in most of the `Service` implementations.  If what we're looking for is cancellation, then we can just create a channel for that purpose.  `context.Context` is meant to traverse through API boundaries and contain request scoped data.  I've seen a lot of examples of go code using them to handle long running processes and I feel like it introduces more weight than required.

In this case the `Service.ctx` is being wrapped to add cancellation to the current `epochHandler`.  So to my knowledge that the `Service.ctx` was created with `context.WithCancel`, then `ctx.Done()` would close if the original `CancelFunc` was called.  But we wrap this context again in line 347.  So does that mean that if we call the local `cancel CancelFunc` than any goroutines listening on `Service.ctx` would receive the channel close? I would assume no, but these are things that make the code more difficult to understand.  I think as a best practice you shouldn't call `context.WithCancel` or any of the other `context.WithXXX` functions on supplied contexts.

I think `epochHandler.run()` should be modified to accept a `context.Context` that's not the original one stored in  `Service`.  This `runEngine` function should listen on the `Service` one and trigger cancellation within this function for the current `epochHandler`.

Reply:
> I'm of the proponent where we should not be using context.Context within package boundaries

The problem is some of the std lib uses context like `http.NewRequestWithContext` so how do you do it then? There is no other way really to cancel an http request otherwise.

Also contexts are used for a few small cases whereas channels can be used for various things. In my opinion it's always clearer to use contexts to cancel long running operations. Channels are useful to return readiness or errors from long running operations though.